application/src/main/java/org/ehrbase/application/abac/CustomMethodSecurityExpressionRoot.java

```diff
@@ -0,0 +1,523 @@
+   * @throws IOException On parsing error
+   * @throws InterruptedException On error while communicating with the ABAC server
+   */
+  public boolean checkAbacPre(String type, Authentication auth, String subject, Object payload,
+      String contentType)
+      throws IOException, InterruptedException {
+
+    // @PreAuthorize will give different types, e.g. String (for composition), EhrStatus,...
+    // so just pipe it through to templateHandling and make by-type handling there
+
+    return checkAbac(type, auth, subject, payload, contentType, PRE);
+  }
+
+  /**
+   * Builds the ABAC request with given data and evaluates the ABAC's response.
+   * @param type Object type of scope
+   * @param auth Authentication stored in the security context
+   * @param subject Subject ID from the current EHR context
+   * @param payload Payload object, either request's input or response's output
+   * @param contentType Content type from the scope
+   * @param authType Pre- or PostAuthorize, determines payload style (string or object)
+   * @return True if ABAC returns a positive feedback, False if not
+   * @throws IOException On parsing error
+   * @throws InterruptedException On error while communicating with the ABAC server
+   */
+  private boolean checkAbac(String type, Authentication auth, String subject, Object payload,
+      String contentType, String authType) throws IOException, InterruptedException {
+    // Set type specific settings:
+    // Extract and set parameters according to which parameters are configured
+    List<String> policyParameters;
+    // Build abac server request, depending on type
+    String requestUrl = abacConfig.getServer().toString();
+    switch (type) {
+      case BaseController.EHR:
+        policyParameters = new ArrayList<>(Arrays.asList(abacConfig.getPolicyEhrParameters()));
+        requestUrl = requestUrl.concat(abacConfig.getPolicyEhrName());
+        break;
+      case BaseController.EHR_STATUS:
+        policyParameters = new ArrayList<>(Arrays.asList(abacConfig.getPolicyEhrStatusParameters()));
+        requestUrl = requestUrl.concat(abacConfig.getPolicyEhrStatusName());
+        break;
+      case BaseController.COMPOSITION:
+        policyParameters = new ArrayList<>(Arrays.asList(abacConfig.getPolicyCompositionParameters()));
+        requestUrl = requestUrl.concat(abacConfig.getPolicyCompositionName());
+        break;
+      case BaseController.CONTRIBUTION:
+        policyParameters = new ArrayList<>(Arrays.asList(abacConfig.getPolicyContributionParameters()));
+        requestUrl = requestUrl.concat(abacConfig.getPolicyContributionName());
+        break;
+      case BaseController.QUERY:
+        policyParameters = new ArrayList<>(Arrays.asList(abacConfig.getPolicyQueryParameters()));
+        requestUrl = requestUrl.concat(abacConfig.getPolicyQueryName());
+        break;
+      default:
+        throw new InternalServerException("ABAC: Invalid type given from Pre- or PostAuthorize");
+    }
```

Comment:
Linked to my comment on `AbacConfig`, I think that using `@ConfigurationProperties`  with a Map<String, PolicyParameters> could be simpler here.

Reply:
I couldn't get rid of this switch structure here, but the `AbacConfig` is much leaner with its own static class und enums, so the setting are type safe. Thanks!